internal/controller/atlasipaccesslist/ipaccesslist.go

```diff
@@ -0,0 +1,152 @@
+
+	if !ipAccessList.GetDeletionTimestamp().IsZero() {
+		if existInAtlas {
+			r.Log.Infof("deleting ip access list from project %s", projectID)
+			return r.deleteIPAccessList(ctx, ipAccessListService, ipAccessList, projectID, atlasIPAccessList, false)
+		}
+
+		r.Log.Info("releasing ip access list resource for deletion")
+		return r.unmanage(ctx, ipAccessList)
+	}
+
+	if !existInAtlas {
+		r.Log.Infof("creating ip access list on project %s", projectID)
+		return r.createIPAccessList(ctx, ipAccessListService, ipAccessList, projectID, akoIPAccessList)
+	}
+
+	r.Log.Infof("managing ip access list on project %s", projectID)
+	return r.manageIPAccessList(ctx, ipAccessListService, ipAccessList, projectID, akoIPAccessList, atlasIPAccessList)
+}
+
+func (r *AtlasIPAccessListReconciler) createIPAccessList(
+	ctx *workflow.Context,
+	ipAccessListService ipaccesslist.IPAccessListService,
+	ipAccessList *akov2.AtlasIPAccessList,
+	projectID string,
+	akoIPAccessList ipaccesslist.IPAccessEntries,
+) ctrl.Result {
+	err := ipAccessListService.Add(ctx.Context, projectID, akoIPAccessList)
+	if err != nil {
+		return r.terminate(ctx, ipAccessList, api.IPAccessListReady, workflow.IPAccessListFailedToCreate, err)
+	}
+
+	return r.watchState(ctx, ipAccessListService, ipAccessList, projectID, akoIPAccessList)
+}
+
+func (r *AtlasIPAccessListReconciler) deleteIPAccessList(
+	ctx *workflow.Context,
+	ipAccessListService ipaccesslist.IPAccessListService,
+	ipAccessList *akov2.AtlasIPAccessList,
+	projectID string,
+	atlasIPAccessList ipaccesslist.IPAccessEntries,
+	partial bool,
+) ctrl.Result {
+	for _, entry := range atlasIPAccessList {
+		err := ipAccessListService.Delete(ctx.Context, projectID, entry)
+		if err != nil {
+			return r.terminate(ctx, ipAccessList, api.IPAccessListReady, workflow.IPAccessListFailedToDelete, err)
+		}
+
+		ctx.EnsureStatusOption(status.RemoveIPAccessListEntryStatus(entry.ID()))
+	}
+
+	if partial {
+		return r.inProgress(ctx, ipAccessList, api.IPAccessListReady, workflow.IPAccessListPending, "Atlas has started to delete access list entries")
+	}
+
+	return r.unmanage(ctx, ipAccessList)
+}
+
+func (r *AtlasIPAccessListReconciler) manageIPAccessList(
+	ctx *workflow.Context,
+	ipAccessListService ipaccesslist.IPAccessListService,
+	ipAccessList *akov2.AtlasIPAccessList,
+	projectID string,
+	akoIPAccessList ipaccesslist.IPAccessEntries,
+	atlasIPAccessList ipaccesslist.IPAccessEntries,
+) ctrl.Result {
+	toAdd := make(ipaccesslist.IPAccessEntries, len(akoIPAccessList))
+	for ID, entry := range akoIPAccessList {
+		if _, ok := atlasIPAccessList[ID]; !ok {
+			r.Log.Debugf("enqueue %s to be added on project %s", entry.ID(), projectID)
+			toAdd[ID] = entry
+		}
+	}
+
+	if len(toAdd) > 0 {
+		return r.createIPAccessList(ctx, ipAccessListService, ipAccessList, projectID, toAdd)
+	}
+
+	toDelete := make(ipaccesslist.IPAccessEntries, len(atlasIPAccessList))
+	for ID, entry := range atlasIPAccessList {
```

Comment:
when inlined with `handleIPAccessList`, move this into a helper function so the state machine handling is less bloated.